enhancements/maintenance/machine-maintenance-operator.md

```diff
@@ -0,0 +1,237 @@
+
+These processes will only hold true for infra and worker roles of machines within the cluster. 
+
+If a scheduled maintenance is detected for a master node, an alert should be raised.
+
+Not every cloud provider has the same kind of maintenances that inherit that same degradation on the cluster.  Some of the features may need to be toggleable per provider. 
+
+Example: (AWS maintenance)[https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/monitoring-instances-status-check_sched.html#types-of-scheduled-events] vs (GCP maintenance)[https://cloud.google.com/compute/docs/storing-retrieving-metadata#maintenanceevents].
+
+
+### Risks and Mitigations
+
+## Design Details
+
+AWS event types:
+* instance-stop
+* instance-retirement
+* instance-reboot
+* system-reboot
+* system-maintenance
+
+### machinemaintenance controller
+The machinemaintenance controller will iterate through machine CRs and reconcile identified mainteances. It will be responsible for first validating the state of the cluster prior to executing anything on a target object. Validating the state of the cluster will include will initially check for only is the cluster upgrading or is a maintenance already being performed. More use-cases can be added as seen fit. 
+
+If the cluster fails validation (for example is upgrading), the controller will requeue the object and process it again according to its `SyncPeriod` which would currently be proposed at 60 minutes. 
+
+After cluster validation, the controller will ascertain if its in a maintenance window where is can execute maintenances (See open question 2). If in the case no maintenance windows are defined, the controller will continue as true. If the maintenance window logic is only applicable in OSD, the operator could validate if its "managed" prior to expecting these resources.
+
+The event type is then sourced from the CR and then is resolved by either deleting a target machine CR (so the machine-api creates a new one) or raising an alert for manual intervention (master maintenance scheduled). 
+
+This is a very UDP type of approach. 
+
+The MMo could also look to store state of its actions in its on machinemaintenance CR that it would create from a machine CR. 
+
+### Example machinemaintenance CR
+
+```
+apiVersion: machinemaintenance.managed.openshift.io/v1alpha1
+kind: MachineMaintenance
+metadata:
+  name: mm-dofinn-20201705-blz22-worker-ap-southeast-2a-984h4
+  namespace: machine-maintenance-operator
+spec:
+  maintenancescheduled: false
+  eventcode: "instance-stop"
+  eventid: "instance-event-01d0903276a5d038c"
+  machineid: "i-04fdc4494938e5c4a"
+  machinelink: "dofinn-20201705-blz22-worker-ap-southeast-2a-984h4"
+status:
+  maintenance: "in-progress"
+```
+
+The MMO would then delete this CR after validating the target machine has been deleted and new one created indicating that the maintenance is completed. 
```

Comment:
would there be any way to know that the maintenance had completed after the CR is deleted?

my concern is that the maintenance would complete and then the deletion happens and due to timing something might miss that signal. in these cases is there some backup or audit trail (perhaps events) ?

Reply:
>would there be any way to know that the maintenance had completed after the CR is deleted?
Make the exact same query to find any scheduled events for target node. If != nil, either the maintenance didnt clear or a new one is now there.